src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -1498,81 +1441,77 @@ std::vector<MergeTreeData::LoadPartResult> MergeTreeData::loadDataPartsFromDisk(
 
     try
     {
-        for (size_t thread = 0; thread < num_threads; ++thread)
+        while (true)
         {
-            pool.scheduleOrThrowOnError([&, thread, thread_group = CurrentThread::getGroup()]
             {
-                SCOPE_EXIT_SAFE(
-                    if (thread_group)
-                        CurrentThread::detachFromGroupIfNotDetached();
-                );
-                if (thread_group)
-                    CurrentThread::attachToGroupIfDetached(thread_group);
+                std::lock_guard lock(part_select_mutex);
+                if (parts_to_load.empty())
+                    break;
+            }
 
-                while (true)
-                {
-                    PartLoadingTree::NodePtr thread_part;
-                    size_t thread_idx = thread;
+            /// Schedule all the tasks we have for now
+            while (true)
+            {
+                PartLoadingTree::NodePtr current_part;
 
-                    {
-                        std::lock_guard lock{part_select_mutex};
+                {
+                    std::lock_guard lock(part_select_mutex);
+                    if (parts_to_load.empty())
+                        break;
```

Comment:
You can get rid of the ugly double while-cycle by moving "Wait for all scheduled tasks + clear futures" from after the `while` to here, instead of a break. And do one more emptiness check to break after waits.

Reply:
Seems like fixed.